kyuubi-ha/src/main/scala/org/apache/kyuubi/ha/client/ServiceDiscovery.scala

```diff
@@ -342,6 +343,31 @@ object ServiceDiscovery extends Logging {
     }
     serviceNode
   }
+
+  def withLock(
+      zkClient: CuratorFramework,
+      lockPath: String,
+      lockTimeout: Long)(f: => Unit): Unit = {
+    var lock: InterProcessSemaphoreMutex = null
+    try {
+      try {
+        lock = new InterProcessSemaphoreMutex(zkClient, ZKPaths.makePath(lockPath, "lock"))
+        lock.acquire(lockTimeout, TimeUnit.MILLISECONDS)
+      } catch {
+        case e: Exception => throw KyuubiSQLException(s"Lock failed on path [$lockPath]", e)
+      }
+      f
+    } finally {
+      try {
+        if (lock != null) {
+          lock.release()
+        }
+      } catch {
+        case _: Exception =>
```

Comment:
What if we fail to release the lock？So it locks forever？

Reply:
Yes, unless we restart `Kyuubi` Server. In this case, the new process acquires the lock after waiting for `HA_ZK_ENGINE_LOCK_TIMEOUT`.

Reply:
I see. Seems we can use `ENGINE_INIT_TIMEOUT` for the lock timeout?

Reply:
Yes, the default value of `HA_ZK_ENGINE_LOCK_TIMEOUT` is the same as `ENGINE_INIT_TIMEOUT`. Using 'ENGINE_INIT_TIMEOUT' should work. Do we need to modify this?

Reply:
yes, let's use `ENGINE_INIT_TIMEOUT ` to make it simple

Reply:
Update

Reply:
thanks